File: src/main/scala/li/cil/oc/common/item/CPU.scala
package li.cil.oc.common.item

import li.cil.oc.common.item.abstracts.SimpleItem
import net.minecraft.item.Item.Properties
import net.minecraftforge.common.extensions.IForgeItem

import java.util
import net.minecraft.entity.player.PlayerEntity
import net.minecraft.item.ItemStack
import net.minecraft.util.ActionResult
import net.minecraft.util.text.ITextComponent
import net.minecraft.world.World

import scala.language.existentials

class CPU(props: Properties, val tier: Int) extends SimpleItem(props) with IForgeItem with interfaces.ItemTier with interfaces.CPULike {
  @Deprecated
  override def getDescriptionId = super.getDescriptionId + tier

  override def cpuTier = tier

  override protected def tooltipName = Option(unlocalizedName)

  override protected def tooltipData = cpuTooltipData

  override protected def tooltipExtended(stack: ItemStack, tooltip: util.List[ITextComponent]) = cpuTooltipExtended(stack, tooltip)

  override def use(stack: ItemStack, world: World, player: PlayerEntity): ActionResult[ItemStack] = cpuUse(stack, world, player)
}

File: src/main/scala/li/cil/oc/common/item/CPU.java
package li.cil.oc.common.item;

import li.cil.oc.common.item.abstracts.SimpleItem;
import li.cil.oc.common.item.interfaces.CPULike;
import li.cil.oc.common.item.interfaces.ItemTier;
import net.minecraft.entity.player.PlayerEntity;
import net.minecraft.item.ItemStack;
import net.minecraft.util.ActionResult;
import net.minecraft.util.text.ITextComponent;
import net.minecraft.world.World;
import net.minecraftforge.common.extensions.IForgeItem;
import scala.Option;

import java.util.List;
import java.util.Optional;

public class CPU extends SimpleItem implements IForgeItem, ItemTier, CPULike {
    private final int tier;

    public CPU(Properties props, int tier) {
        super(props);
        this.tier = tier;
    }

    @Override
    public String getDescriptionId() {
        return super.getDescriptionId() + tier;
    }

    @Override
    public int cpuTier() {
        return tier;
    }

    @Override
    public Optional<String> tooltipName() {
        return Optional.of(unlocalizedName);
    }

    @Override
    public List<Object> tooltipData() {
        return cpuTooltipData();
    }

    @Override
    public void tooltipExtended(ItemStack stack, List<ITextComponent> tooltip) {
        cpuTooltipExtended(stack, tooltip);
    }

    @Override
    public ActionResult<ItemStack> use(ItemStack stack, World world, PlayerEntity player) {
        return cpuUse(stack, world, player);
    }
}
